### PR TITLE
Lookup user via fingerprint before attempting decrypt

### PR DESCRIPTION
**Pull request by jiwalker-usgs**
First pass at #18, thoughts?

### CODE REVIEW

File: R/secrets.R [hunk 1]
```diff
@@ -306,14 +306,14 @@ share_secret_with_key1 <- function(name, email, aeskey, vault) {
...
-    if (!is.null(aes)) return(aes)
-  }
+  file <- get_secret_user_file_for_key(vault, name, key)
+  if (is.null(file)) return(NULL)
+  
+  aes <- tryCatch(
+    unserialize(rsa_decrypt(read_raw(file), key = key)),
+    error = function(e) NULL
+  )
+  if (!is.null(aes)) return(aes)
```
Comment (gaborcsardi) [inline R/secrets.R]: 312-318 can just be 

```r
tryCatch(
...
)
```

File: R/users.R [hunk 2]
```diff
@@ -179,3 +179,23 @@ users_exist <- function(vault, users) {
 on_failure(users_exist) <- function(call, env) {
   paste0("Secret ", deparse(call$users), " do not exist")
 }
+
+#' @importFrom openssl fingerprint read_pubkey
+
+lookup_user <- function(key, vault = NULL) {
```
Comment (gaborcsardi) [inline R/users.R]: No default value needed for `vault`. Also no `find_vault` call is needed. The public API functions find the vault already.

File: R/users.R [hunk 3]
```diff
@@ -179,3 +179,23 @@ users_exist <- function(vault, users) {
...
+  user <- NULL
+  vault <- find_vault(vault)
+  if (is.character(key)) {
```
Comment (gaborcsardi) [inline R/users.R]: Don't need to handle both cases (file or direct key) here. Just handle the case that actually happens.

File: R/users.R [hunk 4]
```diff
@@ -179,3 +179,23 @@ users_exist <- function(vault, users) {
...
+    key <- tryCatch(read_key(key),
+                    error = function(e) NULL)
+    if (is.null(key)) return(NULL)
+  }
+  fp <- fingerprint(key)
+  for (pubkey in dir(file.path(vault, "users"), pattern='\\.pem$')) {
+    pubkeyfile = file.path(vault, 'users', pubkey)
+    if (as.character(fp) == as.character(fingerprint(read_pubkey(pubkeyfile)))) {
+      user <- sub("\\.pem$", "", pubkey)
+    }
```
Comment (gaborcsardi) [inline R/users.R]: You can return from the loop, no need to continue the loop if you already found the user.

File: R/vault.R [hunk 5]
```diff
@@ -135,6 +135,7 @@ get_secret_file <- function(vault, name) {
 }
 
 get_secret_user_file <- function(vault, name, email) {
+  if (is.null(email)) return(NULL)
```
Comment (gaborcsardi) [inline R/vault.R]: Why is this needed?

File: R/users.R (hunk 3 above)
Comment (jiwalker-usgs) [inline R/users.R]: `rsa_decrypt` handles both file or key, but fingerprint only handles a key, so it throws an error on L190 if I don't first read in the key.

File: R/vault.R (hunk 5 above)
Comment (jiwalker-usgs) [inline R/vault.R]: You're right, this is not needed, it was added to propagate NULL through, but isn't necessary.

**Comment by jiwalker-usgs** [comment]
Made requested changes.

File: R/users.R [hunk 6]
```diff
@@ -179,3 +179,22 @@ users_exist <- function(vault, users) {
 on_failure(users_exist) <- function(call, env) {
   paste0("Secret ", deparse(call$users), " do not exist")
 }
+
+#' @importFrom openssl fingerprint read_pubkey
+
+lookup_user <- function(key, vault) {
+  if (is.character(key)) {
+    key <- tryCatch(read_key(key),
+                    error = function(e) NULL)
```
Comment (gaborcsardi) [inline R/users.R]: Can you pls write this in one line, or like this:
```r
key <- tryCatch(
  read_key(key),
  error = function(e) NULL
)
```
To match the style of the existing code. Thanks.

File: R/users.R [hunk 7]
```diff
@@ -179,3 +179,22 @@ users_exist <- function(vault, users) {
...
+    if (is.null(key)) return(NULL)
+  }
+  fp <- fingerprint(key)
+  for (pubkey in dir(file.path(vault, "users"), pattern='\\.pem$')) {
+    pubkeyfile = file.path(vault, 'users', pubkey)
```
Comment (gaborcsardi) [inline R/users.R]: Can you use `<-` instead of `=`? Thanks.

File: R/users.R [hunk 8]
```diff
@@ -179,3 +179,22 @@ users_exist <- function(vault, users) {
...
+    if (as.character(fp) == as.character(fingerprint(read_pubkey(pubkeyfile)))) {
+      user <- sub("\\.pem$", "", pubkey)
+      return(user)
+    }
+  }
+  return(NULL)
```
Comment (gaborcsardi) [inline R/users.R]: Can you pls omit the `return` and just have `NULL`?

**Comment by gaborcsardi** [comment]
Thanks! It looks great! I made some (more) small requests.

File: R/users.R (hunk 7 above)
Comment (gaborcsardi) [inline R/users.R]: Actually, can you also use double quotes?

Comment (gaborcsardi) [inline R/users.R]: Double quotes in both lines? Thanks.

**Comment by jiwalker-usgs** [comment]
Thanks for pointing out the style differences, changes made.

**Comment by gaborcsardi** [comment]
Thanks!
